File: src/structure/java/stack/Stack.java
package structure.java.stack;

public class Stack<E> implements structure.Stack<E> {
    private E[] data;
    private int size;
    public Stack(){
        data = (E[]) new Object[10];
    }

    public boolean isEmpty(){
        return size==0;
    }


    private void resize(int length){
        E[] newData = (E[]) new Object[length];
        for(int i = 0;i<size;i++){
            newData[i] = data[i];
        }
        data = newData;
    }

    public int getSize() {
        return size;
    }

    @Override
    public void push(E element) {
        if(size==data.length) resize(size*2);
        data[size] = element;
        size++;
    }

    public E pop(){
        E temp = data[size-1];
        size--;
        data[size]=null;
        if(size<data.length/4 && data.length/2!=0) resize(data.length/2);
        return temp;
    }

    public E peek(){
        return data[size-1];
    }
}

File: src/structure/kotlin/stack/Stack.kt
package structure.kotlin.stack

import structure.Stack
import kotlin.Array

class Stack<E> : Stack<E> {
    override fun isEmpty(): Boolean {
        return empty
    }

    private var data = arrayOfNulls<Any>(10) as Array<E>
    private var size = 0
    val empty: Boolean
        get() = size == 0

    override fun getSize(): Int = size

    private fun resize(length: Int) {
        val newData = arrayOfNulls<Any>(length) as Array<E>
        for (i in 0 until size) {
            newData[i] = data[i]
        }
        data = newData
    }

    override fun push(element: E) {
        if (size == data.size) resize(size * 2)
        data[size++] = element
    }

    override fun pop(): E {
        val temp = data[--size]
        if (size < data.size / 4 && data.size / 2 != 0) resize(data.size / 2)
        return temp
    }

    override fun peek(): E = data[size - 1]
}